x-pack/plugins/apm/public/utils/timepicker/index.js

```diff
@@ -49,55 +48,41 @@ export function initTimepicker(history, dispatch, callback) {
       });
 
       history.listen(() => {
-        updateRefreshRate(dispatch, timefilter);
+        updateRefreshRate(dispatch);
         globalState.fetch();
       });
-      timefilter.setTime = (from, to) => {
-        timefilter.time.from = moment(from).toISOString();
-        timefilter.time.to = moment(to).toISOString();
-        $scope.$apply();
-      };
       timefilter.enableTimeRangeSelector();
       timefilter.enableAutoRefreshSelector();
-      timefilter.init();
 
-      updateRefreshRate(dispatch, timefilter);
+      updateRefreshRate(dispatch);
 
-      timefilter.on('update', () => dispatch(getAction(timefilter)));
+      $scope.$listen(timefilter, 'timeUpdate', () =>
+        dispatch(updateTimePickerAction())
+      );
```

Comment:
Having `timefilter` as a singleton is super nice! Now I can (almost) get rid of Angular! Now we can avoid `$scope` completely:

```
timefilter.on('timeUpdate', () => dispatch(updateTimePickerAction()));
```

Reply:
The only thing [$scope.$listen](https://github.com/elastic/kibana/blob/master/src/ui/public/listen/listen.js#L34) does is unsubscribed the listener when $scope is destroyed. I think it is still needed in this case so the listener gets unsubscribed.

Reply:
Nifty, didn't think about that side-effect. I'm hoping to move this out of Angular some time in the future but for now it makes sense.